### ISSUE TITLE
Support gnustl for Android

### ISSUE THREAD

**Issue by KindDragon** [conan-io/conan#1252]
https://developer.android.com/ndk/guides/cpp-support.html

**Comment by lasote** [comment]
Hi! What you mean with "support"? Are you missing different settings default values for Android?

**Comment by KindDragon** [comment]
When I call `conan install/test_package` with `-s compiler.libcxx="gnustl"` it return error:
```
ERROR: Invalid setting 'gnustl' is not a valid 'settings.compiler.libcxx' value.
Possible values are ['libstdc++', 'libstdc++11']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"
```

**Comment by forlayo** [comment]
I'm getting in same issue here, are there something to do to avoid this issue? 

I'm using Qt and it works with ndk 10e and gnustl..

**Comment by memsharded** [comment]
Hi Miguel!

Yes, gnustl is not in the default settings.yml values. You can read: http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting

Basically you can add it yourself to settings.yml file (and if you need to configure with your team, you can use ``conan config install`` command)

Then, how you use it for your build, might require some minor changes. For example the built-in ``AutotoolsBuildEnvironment`` helper or the ``cmake`` generator might not work out of the box with it. Fortunately, those helpers are easily adaptable, so please ask if you need help with them.
Thanks!

**Comment by lasote** [comment]
Hi, from the official docs (https://developer.android.com/ndk/guides/cpp-support.html) the gnustl is deprecated from ndk16 and will be removed. 
It's ok if you add it to the `settings.yml` for your company, but we try to keep the `settings.yml` clean/complete enough for the general purpose of the community, so as it is deprecated in the latest version and it is not widely requested, I would keep it out from the default `settings.yml` file. If this issue gains upvotes we will reconsider it, because it would prove the community need.

```
gnustl

The GNU C++ Library is called gnustl on Android to differentiate it from the system runtime. This runtime is the libstdc++ available on a GNU/Linux system.

This runtime is tightly coupled to GCC, which is no longer supported in the NDK. As such, it has not received updates for several releases. The version in the NDK only supports C++11, and some portions of this library are incompatible with Clang.

Note: This library will be deprecated and removed in a future NDK release. Beginning with NDK r16, you should use libc++ instead.
```

**Comment by yowidin** [comment]
Hi folks, on a somehow related note:
From https://developer.android.com/ndk/guides/cpp-support.html:
> The shared library for this runtime is libc++_shared.so, and the static library is libc++_static.a

From https://developer.android.com/ndk/guides/standalone_toolchain.html#creating_the_toolchain:
> Use -static-libstdc++ when using libc++ for the static library version, or -lstdc++ (implicit) to link against the static library version of any other implementation

Currently I couldn't find any way to say conan that I want to use the shared version of the libc++ on Android. I can add appropriate `CXXFLAGS`  in toolchain, but conan will automatically add `-stdlib=libc++` in it's CMake helper which will implicitly link the static version.

As for adapting build helpers, can you please provide some basic directions where such kind of tweaks should be implemented?

**Comment by memsharded** [comment]
I can help with build helpers. They typically have access to information between the creation and the actual calls. So something like:

```python
def build(self):
    cmake = CMake(self)
    if not self.options.shared:
          cmake.definitions["CONAN_LIBCXX"] = "-static-libstdc++"
    cmake.configure()
    ...
```
You can see defined vars here: http://docs.conan.io/en/latest/reference/build_helpers/cmake.html#definitions

Regarding this specific issue about the standard lib, I don't know if @lasote has some further insights, or if it is something worth improving in conan.

**Comment by yowidin** [comment]
Thank you for the response. 

The problem with your approach is that a package itself should be aware of possible library configurations. It seems impractical for me - making all the packages  aware of all the possible platforms, their libstd requirements etc. 

My initial thought was to create a `android-ndk-toolchain` which will have appropriate libstd options (e.g. libstd=['c++_shared', 'c++_static', 'libstd_shared', 'libstd_static']) and set the compiler flags accordingly. 

Currently it seems impossible without adjusting all the libraries recipes that could be built for the Android. If a library recipe uses unmodified CMake helper and just calls `conan_basic_setup()` the libstd settings will be overwritten. And modifying all the recipes is not a whole lot better than throwing out existing settings.yml and redesigning them all from the scratch just for our needs. :)

**Comment by lasote** [comment]
It is very confusing:

> Use -static-libstdc++ when using libc++ for the static library version, or -lstdc++ (implicit) to link against the static library version of **any other implementation**. 

That means that these flags are only to chose the static or the shared of the standard library irrespective of it is GNU libstdc++ STLport or libc++?

First of all I would like to understand the issue, maybe we are missing a subsetting of Android to model it. Right now you could add that flag in the to the profile in CFLAGS and CXXFLAGS, that will affect automatically to all your libraries.

**Comment by yowidin** [comment]
If I understood the issue I had correctly it looks something like this:
The only officially supported libraries from now on are `libc++_shared.so` and `libc++_static.a`. Per default the standalone android compiler will link against the shared library implicitly. There is no `libc++.so` inside the standalone toolchain. 

So, as mentioned in #2049:
> Emphasized the -lc++. I believe the -stdlib=libc++ is causing this flag to appear. The only workaround that works for me right now is to explicitely set CONAN_LIBCXX to an empty string.

Then building for Android regardless the CFLAGS and CXXFLAGS, the `-stdlib=libc++` will be added which will cause compilation failure, since there is no `libc++` in standalone compilers, only `libc++_shared` and `libc++_static` 

> That means that these flags are only to chose the static or the shared of the standard library irrespective of it is GNU libstdc++ STLport or libc++?

Seems so, `-static-libstdc++` or `-lstdc++` will cause the static standard library linking.

**Comment by lasote** [comment]
Right, thanks. And what would happen if you introduce in your libcxx setting in `settings.yml` `libc++_shared` and `libc++_static` and try to specify that setting? it works? Maybe we just need to add them to the list of `clang` compiler.

**Comment by yowidin** [comment]
It would probably fix this particular problem: there would be no match inside the `conan_set_libcxx()` macro so no `-stdlib=libc++` flag will be added. But it look a bit dirty to me - adding hacks to settings file which make no sense for normal Linux/Windows builds.

I personally tend to rewrite the whole `settings.yml` file to suit my needs (include c++ standard, clean up `libcxx`, add shared/static runtime etc.) and then just create dummy recipes which will handle those settings, adding appropriate flags and include those recipes via `[build_requires]` inside a profile generator.

**Comment by memsharded** [comment]
Closing this issue as too old.
All the build system integrations have changed since then, please create updated tickets if necessary.
